File: examples/typegraphs/random-field.ts
import { Policy, t, typegraph } from "@typegraph/sdk/index.js";
import { DenoRuntime } from "@typegraph/sdk/runtimes/deno.js";

typegraph({
  name: "random-field",
  // skip:next-line
  cors: { allowOrigin: ["https://metatype.dev", "http://localhost:3000"] },
}, (g) => {
  const deno = new DenoRuntime();
  const pub = Policy.public();

  const bonusItems = t.list(t.enum_(["+1 gold", "+1 metal"]));
  const dailyBonus = t.struct(
    {
      performance: t.integer(),
      bonus: bonusItems.fromRandom(), // this field is now generated randomly
    },
  );

  // set a custom seed
  g.configureRandomInjection({ seed: 1234 });

  g.expose({
    get_bonus: deno.func(
      dailyBonus,
      t.string(),
      {
        code: ({ performance, bonus }) =>
          `Daily bonus: ${(performance > 100 ? bonus : ["none"]).join(", ")}`,
      },
    ),
  }, pub);
});

File: examples/typegraphs/random-field.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typegraph import Graph, Policy, t, typegraph

# skip:start
from typegraph.graph.params import Cors
from typegraph.runtimes.deno import DenoRuntime

# skip:end


@typegraph(
    # skip:start
    cors=Cors(allow_origin=["https://metatype.dev", "http://localhost:3000"]),
    name="random-field",
    # skip:end
)
def random_field(g: Graph):
    deno = DenoRuntime()
    pub = Policy.public()

    bonus_items = t.list(t.enum(["+1 gold", "+1 metal"]))
    daily_bonus = t.struct(
        {
            "performance": t.integer(),
            "bonus": bonus_items.from_random(),  # this field is now generated randomly
        },
    )

    # set a custom seed
    g.configure_random_injection(seed=1234)

    g.expose(
        pub,
        get_bonus=deno.func(
            daily_bonus,
            t.string(),
            code="""({ performance, bonus }) => `Daily bonus: ${
                (performance > 100 ? bonus : ['none']).join(', ')
            }`;
            """,
        ),
    )
